### PR TITLE
Adding crosstargeting in order to be able to call WinRT APIs in net5.0

### PR DESCRIPTION
**Pull request by joperezr**
Fixes https://github.com/dotnet/iot/issues/1091
Fixes https://github.com/dotnet/iot/issues/1103
Fixes https://github.com/dotnet/iot/issues/1193

cc: @richlander @pgrawehr @Ellerbach @krwq @ericstj @terrajobst @jkoritzinsky @dsplaisted 

This PR will add cross-targeting to System.Device.Gpio package so that it will now have 3 different assets in the package:
- One netstandard2.0 asset: This will be used by consuming apps that are targeting netstandard2.0 compatible frameworks **that are not net5.0**. Nothing has changed from this asset from what we shipped before.
- One net5.0 asset: This asset will be for people targeting net5.0 that **don't intend to target Windows**. This implementation is very similar to the netstandard2.0 one, except that it will now throw platform not supported exceptions when running on Windows and trying to use types that depend on WinRT APIs.
- One net5.0-windows10.0.17763.0 asset: This asset will be for people targeting net5.0 that **want to run in Windows**. It will be using the new CSWinRT project to call into WinRT APIs and don't depend on System.Runtime.* package any longer. In order for folks to get this asset, they will need to change the target frameworks section in their projects and add `net5.0-windows10.0.17763.0` so that the package will work at runtime. **This is a breaking change**

As explained above, when running on net5.0 and targeting Windows it will be required to edit the target frameworks section of the project. This is the only way to get this working for now, and we are looking into some embedding scenarios (once CSWinRT supports them) so that we can change the way we distribute our package so that consuming projects won't require to cross-target. As part of this change, we are also adding a targets file with the package that will try to detect cases where cross-targeting is required, in order to catch those cases early and give helpful messages to consumers so that they can change their project accordingly. Here is how the package will look now:
image

And here are the package dependencies after this change:
image

Any feedback is appreciated 😄

### CODE REVIEW

File: src/System.Device.Gpio/build/net5.0/System.Device.Gpio.targets
Comment (ericstj) [inline src/System.Device.Gpio/build/net5.0/System.Device.Gpio.targets:4]: You only want to cause runnable applications to fail, not libraries.  It's ok for another library to target just `net5.0`.  Can you make sure this check only applies to apps?  Also it's best practice to have a way to disable such errors.  Consider adding a property to suppress this.

**Review by jkoritzinsky — APPROVED** [review]

File: src/System.Device.Gpio/System/Device/Gpio/Drivers/Windows10Driver.notSupported.cs
Comment (ericstj) [inline src/System.Device.Gpio/System/Device/Gpio/Drivers/Windows10Driver.notSupported.cs:22]: Can you factor this to a common resource string with parameter (`{0}`) for type name to avoid duplication?  Perhaps you might also want to parameterize that windows version?  Not sure how you plan to keep that in sync with TargetFramework.

File: src/System.Device.Gpio/build/net5.0/System.Device.Gpio.targets
Comment (dsplaisted) [inline src/System.Device.Gpio/build/net5.0/System.Device.Gpio.targets:5]: It doesn't have to be exactly 17763, it could be that version or higher.  Also, I'm not sure if the RuntimeIdentifier will necessarily be set in all scenarios where you'd want this check.

Comment (joperezr) [reply src/System.Device.Gpio/build/net5.0/System.Device.Gpio.targets:5]: I'll change the message to say that 17763 is just the lower bound. What would you recommend instead of using RuntimeIdentifier? I know that we might still have some cases where we won't be able to catch problems, like for example shared framework applications, but that's why for those we are relying on the PlatformNotSupported exception messages to help figure out how to fix the targeting. Of course, if there are better ways to catch it at compile time instead of runtime that would be much prefered.

**Comment by joperezr** [comment]
@dsplaisted looking at the CI failures seemed like the problem is that the net5.0-windows build was failing on Linux and OSX as apparently, you can only build for that TFM with a Windows SDK. Is that by design or will support for targeting windows from a non-windows machine will be added eventually? I just added a change that disables Windows-specific build on non-Windows but this scenario seems kind of busted as in theory you would only need the Windows targeting pack and we have supported in the past targeting TFMs that might not necessarily run on the machine you are in, this also means for us that our Linux and OSX contributors won't be able to make changes and build windows specific files.

File: src/System.Device.Gpio/System/Device/Gpio/Drivers/Windows10Driver.notSupported.cs
Comment (joperezr) [reply src/System.Device.Gpio/System/Device/Gpio/Drivers/Windows10Driver.notSupported.cs:22]: The repo doesn't currently use ResourceManager and it is in our plans to add that support, but ideally, when added I would rather do it to the whole library and not just this string. I can centralize the format of this string for now as some sort of constant format, and then for vNext work on moving everything into resources instead, does that sound good?

File: src/System.Device.Gpio/build/net5.0/System.Device.Gpio.targets
Comment (dsplaisted) [reply src/System.Device.Gpio/build/net5.0/System.Device.Gpio.targets:4]: Good point about not generating the error for libraries.  Relatedly, you probably want to use the `buildTransitive` folder instead of the `build` folder so that if an app transitively references this package it will get the build error.

Comment (ericstj) [reply src/System.Device.Gpio/build/net5.0/System.Device.Gpio.targets:4]: Also per offline conversation you might be able to use `AppHostRuntimeIdentifier` to catch the case where someone is building a framework-dependent app on Windows.

**Comment by dsplaisted** [comment]
> @dsplaisted looking at the CI failures seemed like the problem is that the net5.0-windows build was failing on Linux and OSX as apparently, you can only build for that TFM with a Windows SDK. Is that by design or will support for targeting windows from a non-windows machine will be added eventually? I just added a change that disables Windows-specific build on non-Windows but this scenario seems kind of busted as in theory you would only need the Windows targeting pack and we have supported in the past targeting TFMs that might not necessarily run on the machine you are in, this also means for us that our Linux and OSX contributors won't be able to make changes and build windows specific files.

It's by design for now.  Hopefully when we have optional workload support we'll eventually make the Windows optional workload work on other operating systems.

If you just want to use the types, you may be able to add an `Update` statement to set the `IsWindowsOnly` metadata to false for the `KnownFrameworkReference` items for Windows.

File: src/System.Device.Gpio/System/Device/Gpio/Drivers/Windows10Driver.notSupported.cs
Comment (ericstj) [reply src/System.Device.Gpio/System/Device/Gpio/Drivers/Windows10Driver.notSupported.cs:22]: Even though you aren't using ResX here you can still de-dup the string to a single helper method.

Comment (joperezr) [reply src/System.Device.Gpio/System/Device/Gpio/Drivers/Windows10Driver.notSupported.cs:22]: yeah, that's what I meant with centralizing the format. I'll do that then.

**Comment by joperezr** [comment]
> If you just want to use the types, you may be able to add an Update statement to set the IsWindowsOnly metadata to false for the KnownFrameworkReference items for Windows.

I actually expect very few people to actually want to change things on the Windows-specific build for now, so I'm fine with just not building it on Linux and OSX at the moment, but my 2 cents here are just that it would be great to either document the `isWindowsOnly` metadata workaround for other folks or to allow people to target different platform than the one they run in.

File: src/System.Device.Gpio/build/net5.0/System.Device.Gpio.targets
Comment (joperezr) [reply src/System.Device.Gpio/build/net5.0/System.Device.Gpio.targets:4]: If I do that, then folks creating framework-dependent apps on Windows with the intention of running them on Linux will get the error, which I actually believe will be one of the most common scenarios. Would it be ok to not protect against that case and instead let the runtime exception catch such scenarios given they should, in theory, be the least common case?

File: src/System.Device.Gpio/System/Device/Gpio/Drivers/Windows10Driver.notSupported.cs
Comment (ericstj) [reply src/System.Device.Gpio/System/Device/Gpio/Drivers/Windows10Driver.notSupported.cs:22]: Yep, sorry I had replied here before seeing your reply.  GH didn't refresh for some reason.

File: src/System.Device.Gpio/build/net5.0/System.Device.Gpio.targets
Comment (ericstj) [reply src/System.Device.Gpio/build/net5.0/System.Device.Gpio.targets:4]: > creating framework-dependent apps on Windows with the intention of running them on Linux

We could teach these folks to use `AppHostRuntimeIdentifier` but maybe that's overkill.  I'm fine with whatever you think is best here.  We could even forgo the whole targets if we don't think it'll help enough to warrant the added complexity.

Comment (joperezr) [reply src/System.Device.Gpio/build/net5.0/System.Device.Gpio.targets]: I think that as long as it catches some of the cases it is worth keeping, as a build time error is always preferred over runtime error, but just want to make sure that we don't catch false-positives by making our safety net too general.

File: src/System.Device.Gpio/build/net5.0/System.Device.Gpio.targets
Comment (pgrawehr) [reply src/System.Device.Gpio/build/net5.0/System.Device.Gpio.targets]: Not sure this should cause an error at all. Even if I'm running on windows, I might not need to use the windows specific implementation, especially if I'm not using the WindowsGpioDriver (but maybe only an Arduino driver).

Comment (joperezr) [reply src/System.Device.Gpio/build/net5.0/System.Device.Gpio.targets]: While most (if not all) of this library would render useless if we can call WinRT in Windows, I do agree with you that perhaps Error is too aggressive even though it is suppressable. @ericstj do you think switching to Warning would be appropriate here?

**Comment by pgrawehr** [comment]
I do want to fully understand what's going on here (so far I honestly don't get it yet). Please do not merge before I can take an adequate review.

**Comment by joperezr** [comment]
@pgrawehr see the description where I added info on the three assets distributed by the package and when would each apply. In there I provide a high level description of what is changing with this PR, but if you have any follow up questions, I’m happy to answer.

File: src/System.Device.Gpio/System.Device.Gpio.csproj
Comment (pgrawehr) [inline src/System.Device.Gpio/System.Device.Gpio.csproj:52]: I think the PR lacks files. This directory is empty for me (except for "Interop.WaitForCompletion.cs", which only contains a helper method). 

Besides, I don't understand why "net5.0" or not would have anything to do with Windows or not.

Comment (joperezr) [reply src/System.Device.Gpio/System.Device.Gpio.csproj:52]: > I think the PR lacks files. This directory is empty for me (except for "Interop.WaitForCompletion.cs", which only contains a helper method).

I know this is the only WinRT file today, but I used wildcards as any file that goes in it in the future would have to be removed, and because everything is added to all configurations by default, I wanted to make sure that we guarded against new files or renames.

> Besides, I don't understand why "net5.0" or not would have anything to do with Windows or not.

Our net5.0 configuration is effectively a non-Windows build since we want everyone targeting net5.0 and Windows to use the net5.0-windows version instead. That is why net5.0 version now throws anytime you are trying to use anything that relies on WinRT code.

File: src/System.Device.Gpio/build/net5.0/System.Device.Gpio.targets
Comment (joperezr) [reply src/System.Device.Gpio/build/net5.0/System.Device.Gpio.targets]: Chatted offline, I'll keep as error for now since we want to make sure that users won't miss it, but keep it suppressible so that users can disable if they think they are a false positive.

**Comment by joperezr** [comment]
I'll merge this now to unblock progress on getting the repo ready for shipping our 1.1 release but feel free to let me know if there are any additional comments/questions around this change and I'm happy to address/answer them.

**Pull request merged**
